code/trans.c

```diff
@@ -0,0 +1,377 @@
+  AVERT(Transform, transform);
+  ControlFree(transform->arena, p, size);
+}
+
+
+Res TransformCreate(Transform *transformReturn, Arena arena)
+{
+  Transform transform;
+  Res res;
+  void *p;
+
+  AVER(transformReturn != NULL);
+  AVERT(Arena, arena);
+
+  res = ControlAlloc(&p, arena, sizeof(TransformStruct));
+  if (res != ResOK)
+    goto failAlloc;
+  transform = (Transform)p;
+  
+  transform->oldToNew = NULL;
+  transform->arena = arena;
+  transform->epoch = ArenaEpoch(arena);
+  transform->aborted = FALSE;
+
+  transform->sig = TransformSig;
+
+  AVERT(Transform, transform);
+
+  res = TableCreate(&transform->oldToNew,
+                    0, /* don't grow table until TransformAddOldNew */
+                    transformTableAlloc,
+                    transformTableFree,
+                    transform,
+                    0, 1); /* these can't be old references */
+  if (res != ResOK)
+    goto failTable;
+
+  *transformReturn = transform;
+  return ResOK;
+
+failTable:
+  ControlFree(arena, transform, sizeof(TransformStruct));
+failAlloc:
+  return res;
+}
+
+
+void TransformDestroy(Transform transform)
+{
+  Arena arena;
+  Table oldToNew;
+
+  AVERT(Transform, transform);
+
+  /* TODO: Log some transform statistics. */
+
+  /* Workaround bootstrap problem, see .check.boot */
+  oldToNew = transform->oldToNew;
+  transform->oldToNew = NULL;
+  TableDestroy(oldToNew);
+
+  arena = TransformArena(transform);
+  transform->sig = SigInvalid;
+  ControlFree(arena, transform, sizeof(TransformStruct));
+}
+
+
+/* TransformArena -- return transform's arena
+ * 
+ * Must be thread-safe as it is called outside the arena lock. See
+ * <design/thread-safety/#sol.check>
+ */
+
+Arena TransformArena(Transform transform)
+{
+  Arena arena;
+  AVER(TESTT(Transform, transform));
+  arena = transform->arena;
+  AVER(TESTT(Arena, arena));
+  return arena;
+}
+
+
+Res TransformAddOldNew(Transform transform,
+                       Ref old_list[],
+                       Ref new_list[],
+                       Count count)
+{
+  Res res;
+  Index i;
+  Count added = 0;
+
+  AVERT(Transform, transform);
+  AVER(old_list != NULL);
+  AVER(new_list != NULL);
+  /* count: cannot check */
+  
+  res = TableGrow(transform->oldToNew, count);
+  if (res != ResOK)
+    return res;
+
+  for(i = 0; i < count; ++i) {
+    /* NOTE: If the mutator isn't adding references while the arena is parked,
+       we might need to access the client-provided lists, using ArenaRead. */
+    if(old_list[i] == NULL)
+      continue;  /* permitted, but no transform to do */
+    if(old_list[i] == new_list[i])
+      continue;  /* ignore identity-transforms */
+
+    /* Old refs must be in managed memory. */
+    {
+      Seg seg;
+      AVER(SegOfAddr(&seg, transform->arena, old_list[i]));
+    }
+
+    res = TableDefine(transform->oldToNew, (Word)old_list[i], new_list[i]);
+    AVER(res != ResFAIL); /* It's a static error to add the same old twice. */
```

Comment:
M. This restriction is not documented and may be unnecessary.  rule.generic.achieve, rule.generic.complete

Reply:
Reject: The restriction is documented, and it's probably good to keep because it prevents programmers accidentally saying they want to transform A to B and then A to C.